### PR TITLE
[feat] create settings page (auth)

### PR DESCRIPTION
**Pull request by ccatherinetan**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### :palm_tree: What's new in this PR
[//]: # "REQUIRED - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
Created a settings page with SIGN OUT and RESET PASSWORD. 

### :evergreen_tree: Screenshots
[//]: # "REQUIRED for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! If you are making changes here, please CC: @davidqing6432 at the bottom."

## :deciduous_tree: How to review
[//]: # 'REQUIRED - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'
**LOGIN** 
- go to http://localhost:3000/login to log in (or create a new account if you don't already have one) 
- this should reroute you to the home page 
- check that you're logged in on supabase → auth → check time of last sign in https://supabase.com/dashboard/project/grffggxilkypzxwecaxz/auth/users
- go to http://localhost:3000/settings → you should see a RESET PASSWORD button and a SIGN OUT button 

**RESET PASSWORD** 

- click Reset Password button → should reroute you to http://localhost:3000/reset-password  
- enter your new password and hit Reset Password → should reroute you to the LOGIN page 
- NOTE: resetting the password should automatically sign you out 

**SIGN OUT** (but first, LOGIN again, otherwise you won't be able to sign out) 
- go to http://localhost:3000/settings → you should see a RESET PASSWORD button and a SIGN OUT button 
- click SIGN OUT → should reroute you to the LOGIN page
- you will be signed out (no direct way to check this, but if you try to sign out when already signed out, you'll get an error)

## :seedling: Next steps 
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

### :information_source: Online sources
[//]: # 'Optional - copy links to any tutorials or documentation that was useful to you when working on this PR'

### :potted_plant: Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

[//]: # 'This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
CC: @pragyakallanagoudar

### CODE REVIEW

File: src/app/settings/page.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import { useRouter } from 'next/navigation';
+import supabase from '../../api/supabase/createClient';
+
+export default function Settings() {
+    const { push } = useRouter();
+    const signOut = async () => {
+        const { error } = await supabase.auth.signOut()
+        if (error) {
+            throw new Error(`An error occurred trying to sign out: ${error}`);
```
Comment (jinkang-0) [inline src/app/settings/page.tsx]: Minor edit: I think the error object that will be returned can just be thrown immediately (i.e. `throw error`), no need to make a new Error object.

File: src/app/login/page.tsx [hunk 2]
```diff
@@ -1,35 +1,35 @@
...
 
     if (error) {
       throw new Error(`An error occurred trying to sign up: ${error}`);
     }
-
-    return data;
+    push("/");
   };
 
   const signInWithEmail = async () => {
```
Comment (sauhardjain) [inline src/app/login/page.tsx]: Rename to `handleSignIn` for naming consistency with other event handlers.

File: src/app/reset-password/page.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import { useRouter } from 'next/navigation';
+import supabase from '../../api/supabase/createClient';
+
+export default function ResetPassword() {
+    const [newPassword, setPassword] = useState('');
+    const { push } = useRouter();
+    const resetPassword = async () => {
+        await supabase.auth.updateUser({ password: newPassword});
+        push("/login");
+    }
```
Comment (sauhardjain) [inline src/app/reset-password/page.tsx]: Add error handling similar to how you did for `handleSignUp`.

File: src/app/settings/page.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            <button type="button" onClick={signOut}>Sign out</button>
+            <button type="button" onClick={resetPassword}>Reset Password</button>
+        </>
+    );
+}
+
+/*
+Send email for password reset using .resetPasswordForEmail while providing a redirectTo parameter
+Email link will work as a magic link and log the user in then take them to the url specified in the redirectTo parameter
+Create form to update the password and call the .updateUser method with the new password */
```
Comment (sauhardjain) [inline src/app/settings/page.tsx]: Not sure if the comments are for TODOs, but we can get rid of them otherwise?

File: src/app/settings/page.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        push("/login");
+    } 
+    const resetPassword = () => {
+        push("/reset-password");
+    }
+
+    return (
+        <>
+            <button type="button" onClick={signOut}>Sign out</button>
+            <button type="button" onClick={resetPassword}>Reset Password</button>
```
Comment (sauhardjain) [inline src/app/settings/page.tsx]: We can just do `onClick={() => push('/reset-password')}` as the `resetPassword` handler is used only for routing and doesn't actually help reset the password.

File: src/app/settings/page.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
+'use client'
+
+import { useRouter } from 'next/navigation';
+import supabase from '../../api/supabase/createClient';
+
+export default function Settings() {
+    const { push } = useRouter();
+    const signOut = async () => {
```
Comment (sauhardjain) [inline src/app/settings/page.tsx]: Rename to `handleSignOut`

File: src/app/reset-password/page.tsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
+"use client"
+
+import { useState } from 'react';
+import { useRouter } from 'next/navigation';
+import supabase from '../../api/supabase/createClient';
+
+export default function ResetPassword() {
+    const [newPassword, setPassword] = useState('');
```
Comment (sauhardjain) [inline src/app/reset-password/page.tsx]: Pretty minor but I'd recommend changing this to `const [password, setPassword] = useState('');` to follow useState naming convention.

**Review by sauhardjain — CHANGES_REQUESTED** [review]
Amazing work @ccatherinetan on setting up auth and routing, you're killing it! Left a few comments here and there to help make the code cleaner.

@pragyakallanagoudar leaving the verification of auth on supabase to you since I'm not added to the org. Next steps would be better error handling, user feedback, more but I'm sure you have it planned.

File: src/app/login/page.tsx (hunk 2 above)
Comment (ccatherinetan) [reply src/app/login/page.tsx]: gotcha!

File: src/app/reset-password/page.tsx (hunk 3 above)
Comment (ccatherinetan) [reply src/app/reset-password/page.tsx]: gotcha!

File: src/app/settings/page.tsx (hunk 5 above)
Comment (ccatherinetan) [reply src/app/settings/page.tsx]: @pragyakallanagoudar in the future, are we going to use a more advanced reset-password involving email redirect link? if so, do we want to keep this function?

File: src/app/settings/page.tsx (hunk 6 above)
Comment (ccatherinetan) [reply src/app/settings/page.tsx]: gotcha!

File: src/app/settings/page.tsx (hunk 5 above)
Comment (pragyakallanagoudar) [reply src/app/settings/page.tsx]: To allow for this possibility, we'll leave the handler function for right now!

**Review by pragyakallanagoudar — APPROVED** [review]

File: src/app/login/page.tsx [hunk 8]
```diff
@@ -1,35 +1,35 @@
...
-    const { data, error } = await supabase.auth.signInWithPassword({
+    const { error } = await supabase.auth.signInWithPassword({
       email,
       password,
     });
 
     if (error) {
       throw new Error(`An error occurred trying to sign in: ${error}`);
```
Comment (jinkang-0) [inline src/app/login/page.tsx]: As it is right now, if an error occurs, it will log "An error occurred trying to sign in: `[object Object]`", since the `error` variable returned by the Supabase call is an error object. I would recommend to just `throw error` instead, or say `error.message` if we want the context.

File: src/app/login/page.tsx [hunk 9]
```diff
@@ -1,35 +1,35 @@
...
   const handleSignUp = async () => {
-    const { data, error } = await supabase.auth.signUp({
+    const { error } = await supabase.auth.signUp({
       email,
       password,
     });
 
     if (error) {
       throw new Error(`An error occurred trying to sign up: ${error}`);
     }
```
Comment (jinkang-0) [inline src/app/login/page.tsx:19]: flagging this for `handleSignUp` as well

File: src/app/reset-password/page.tsx [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import supabase from '../../api/supabase/createClient';
+
+export default function ResetPassword() {
+  const [password, setPassword] = useState('');
+  const { push } = useRouter();
+  const resetPassword = async () => {
+    const { error } = await supabase.auth.updateUser({ password });
+    if (error) {
+      throw new Error(`An error occurred trying to sign in: ${error}`);
+    }
```
Comment (jinkang-0) [inline src/app/reset-password/page.tsx:16]: Flagging the error thing here as well for the reset password page

File: src/app/settings/page.tsx [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+import { useRouter } from 'next/navigation';
+import supabase from '../../api/supabase/createClient';
+
+export default function Settings() {
+  const { push } = useRouter();
+  const handleSignOut = async () => {
+    const { error } = await supabase.auth.signOut();
+    if (error) {
+      throw new Error(`An error occurred trying to sign out: ${error}`);
+    }
```
Comment (jinkang-0) [inline src/app/settings/page.tsx:12]: Flagging the error thing here for Settings page

**Review by jinkang-0 — CHANGES_REQUESTED** [review]
Minor edit to errors

File: src/app/login/page.tsx (hunk 8 above)
Comment (ccatherinetan) [reply src/app/login/page.tsx]: got it! thanks for the tip! i've corrected them all to use `error.message`

File: src/app/settings/page.tsx [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  };
+
+  return (
+    <>
+      <button type="button" onClick={handleSignOut}>
+        Sign out
+      </button>
+      <button type="button" onClick={resetPassword}>
+        Reset Password
+      </button>
```
Comment (jinkang-0) [inline src/app/settings/page.tsx:26]: Could turn this into a regular old <a> tag if we don't need it to do anything else other than redirecting

**Review by jinkang-0 — APPROVED** [review]
lgtm!
